lib/ain-grpc/src/subscription/eth.rs

```diff
@@ -43,148 +48,159 @@ impl MetachainPubSubServer for MetachainPubSubModule {
-                            debug!(target: "pubsub",
-                                "Received transaction hash in newPendingTransactions: {:x?}",
-                                hash
-                            );
-                            let _ = sink.send(&PubSubResult::TransactionHash(hash));
+                    debug!("Ws connection ended, thread closing");
+                    Ok::<(), anyhow::Error>(())
+                };
+                self.tokio_runtime.spawn(fut);
+            }
+            Subscription::NewPendingTransactions => {
+                let fut = async move {
+                    while !sink.is_closed() {
+                        if let Notification::Transaction(hash) = rx.recv().await? {
+                            if !sink.send(&PubSubResult::TransactionHash(hash))? {
+                                break;
+                            }
                         }
                     }
-                }
-                Subscription::Syncing => {
-                    let is_syncing = || -> Result<PubSubSyncStatus, anyhow::Error> {
-                        match ain_cpp_imports::get_sync_status() {
-                            Ok((current, highest)) => {
-                                if current != highest {
-                                    // Convert to EVM block number
-                                    let current_block = handler
-                                        .storage
-                                        .get_latest_block()?
-                                        .ok_or(format_err!("Unable to find latest block"))?
-                                        .header
-                                        .number;
-                                    let starting_block = handler.block.get_starting_block_number();
-                                    let highest_block = current_block + (highest - current); // safe since current can never be more than highest
+                    debug!("Ws connection ended, thread closing");
+                    Ok::<(), anyhow::Error>(())
+                };
+                self.tokio_runtime.spawn(fut);
+            }
+            Subscription::Syncing => {
+                let fut = async move {
+                    let get_sync_status = || -> Result<SyncStatus, anyhow::Error> {
+                        let (current, highest) = ain_cpp_imports::get_sync_status()
+                            .map_err(|_| format_err!("failed to get sync status"))?;
+                        let current_block = handler
+                            .storage
+                            .get_latest_block()?
+                            .ok_or(format_err!("Unable to find latest block"))?
+                            .header
+                            .number;
+                        let diff = U256::from(highest.saturating_sub(current));
+                        let highest_block = current_block.saturating_add(diff);
```

Comment:
Shouldn't we use highest directly here ?

Reply:
Nope, highest refers to the highest DVM block height, while we are trying to get the highest EVM block height that the node is trying to sync to.